### PR TITLE
NIFI-5318 Implement NiFi test harness: initial commit of nifi-test

### PR DESCRIPTION
**Pull request by peter-gergely-horvath**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ x Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
At the moment, the build fails with unit **test failures unrelated to this commit**:
```
[INFO] Running org.apache.nifi.csv.TestCSVValidators
[INFO] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.018 s - in org.apache.nifi.csv.TestCSVValidators
[INFO] Running org.apache.nifi.csv.TestJacksonCSVRecordReader
[ERROR] Tests run: 11, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.033 s <<< FAILURE! - in org.apache.nifi.csv.TestJacksonCSVRecordReader
[ERROR] testUTF8(org.apache.nifi.csv.TestJacksonCSVRecordReader)  Time elapsed: 0.002 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[黃凱揚]> but was:<[???]>
	at org.apache.nifi.csv.TestJacksonCSVRecordReader.testUTF8(TestJacksonCSVRecordReader.java:81)

[INFO] Running org.apache.nifi.csv.TestCSVHeaderSchemaStrategy
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.004 s - in org.apache.nifi.csv.TestCSVHeaderSchemaStrategy
[INFO] Running org.apache.nifi.csv.TestCSVRecordReader
[INFO] Tests run: 21, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.015 s - in org.apache.nifi.csv.TestCSVRecordReader
[INFO] Running org.apache.nifi.csv.TestWriteCSVResult
[INFO] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.004 s - in org.apache.nifi.csv.TestWriteCSVResult
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TestJacksonCSVRecordReader.testUTF8:81 expected:<[黃凱揚]> but was:<[???]>
[ERROR]   TestXMLReader.testAttributePrefix:133 expected:<... ATTR_ID=P3, NAME=Am[é]lie Bonfils, AGE=74}...> but was:<... ATTR_ID=P3, NAME=Am[?]lie Bonfils, AGE=74}...>
[ERROR]   TestXMLReader.testContentField:156 expected:<...apRecord[{CONTENT=Am[é]lie Bonfils, ATTR=at...> but was:<...apRecord[{CONTENT=Am[?]lie Bonfils, ATTR=at...>
[INFO] 
[ERROR] Tests run: 221, Failures: 3, Errors: 0, Skipped: 2
[INFO] 
[INFO] ------------------------------------------------------------------------
```

- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by peter-gergely-horvath** [comment]
@MikeThomsen @joewitt please let me know if you need anything else from my end.

**Comment by MikeThomsen** [comment]
@peter-gergely-horvath looks like you have a stylecheck failure. Did you build the maven module(s) you added using `mvn clean install -Pcontrib-check`? Looks like you have at least a few files that aren't licensed. Depending on what they are, you'll need to either add the license header or just exclude them from the RAT plugin's checks.

**Comment by peter-gergely-horvath** [comment]
@MikeThomsen I tried, but it kept failing with other tests cases, unrelated to my changes, see above in pull request: 
```
At the moment, the build fails with unit test failures unrelated to this commit:
[INFO] Running org.apache.nifi.csv.TestCSVValidators
[INFO] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.018 s - in org.apache.nifi.csv.TestCSVValidators
[INFO] Running org.apache.nifi.csv.TestJacksonCSVRecordReader
[ERROR] Tests run: 11, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.033 s <<< FAILURE! - in org.apache.nifi.csv.TestJacksonCSVRecordReader
[ERROR] testUTF8(org.apache.nifi.csv.TestJacksonCSVRecordReader)  Time elapsed: 0.002 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[黃凱揚]> but was:<[???]>
	at org.apache.nifi.csv.TestJacksonCSVRecordReader.testUTF8(TestJ
```

The fact I cannot get a working build from original/master is kind of painful, do you see any way for me to remedy this?

**Comment by peter-gergely-horvath** [comment]
OK, @MikeThomsen I see some further issues here: I'll close the pull request and try to rectify them.

**Comment by MikeThomsen** [comment]
You can ping me when you're ready.

File: nifi-test/pom.xml [hunk 1]
```diff
@@ -0,0 +1,145 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>nifi</artifactId>
+        <groupId>org.apache.nifi</groupId>
+        <version>1.8.0-SNAPSHOT</version>
+    </parent>
+
+    <modelVersion>4.0.0</modelVersion>
+
```
Comment (ottobackwards) [inline nifi-test/pom.xml]: I don't think this should be thingbiganalytics

File: nifi-test/pom.xml [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    <groupId>com.thinkbiganalytics</groupId>
+    <artifactId>nifi-test</artifactId>
```
Comment (ottobackwards) [inline nifi-test/pom.xml]: nifi-test seems too generic for this.  why not nifi-test-harness ?

File: nifi-test/pom.xml (hunk 1 above)
Comment (MikeThomsen) [reply nifi-test/pom.xml]: Agreed. AFAIK, things like this need to be refactored, where possible, to `org.apache.nifi.*`

File: nifi-test/pom.xml (hunk 2 above)
Comment (MikeThomsen) [reply nifi-test/pom.xml]: +1

File: nifi-docs/src/main/asciidoc/developer-guide.adoc [hunk 3]
```diff
@@ -2296,6 +2296,32 @@ threads that should be used to run the Processor can
...
+<dependency>
+	<groupId>org.apache.nifi</groupId>
+	<artifactId>nifi-test</artifactId>
+	<version>${nifi version}</version>
+</dependency>
+----
+
+For further documentation, please consult the JavaDoc of
+`org.apache.nifi.test.TestNiFiInstance`. For samples, please take a look at
+link:https://github.com/apache/nifi/tree/master/nifi-test/src/test/java/org/apache/test/samples[samples on GitHub^].
```
Comment (joewitt) [inline nifi-docs/src/main/asciidoc/developer-guide.adoc]: what is this link really meant to be?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-documentation/src/main/java/org/apache/nifi/documentation/DocGenerator.java [hunk 4]
```diff
@@ -57,6 +57,9 @@
      * @param extensionMapping extension mapping
      */
     public static void generate(final NiFiProperties properties, final ExtensionMapping extensionMapping) {
+
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-documentation/src/main/java/org/apache/nifi/documentation/DocGenerator.java]: these extra lines should be removed

File: nifi-toolkit/nifi-toolkit-tls/src/main/java/org/apache/nifi/test/SimpleNiFiFlowDefinitionEditor.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+ *
+ * <p>
+ * <strong>CAUTION: THIS IS AN EXPERIMENTAL API: EXPECT CHANGES!</strong>
+ * Efforts will be made to retain backwards API compatibility, but
+ * no guarantee is given.
+ * </p>
+ *
+ * @see TestNiFiInstance.Builder#modifyFlowXmlBeforeInstalling(FlowFileEditorCallback)
+ *
+ * @author Peter G. Horvath
```
Comment (joewitt) [inline nifi-toolkit/nifi-toolkit-tls/src/main/java/org/apache/nifi/test/SimpleNiFiFlowDefinitionEditor.java]: please remove all author tags from all javadocs

File: nifi-toolkit/nifi-toolkit-tls/src/main/java/org/apache/nifi/test/util/FileUtils.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.*;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.Arrays;
+
+public final class FileUtils {
+
+
+    private static final String MAC_DS_STORE_NAME = ".DS_Store";
```
Comment (joewitt) [inline nifi-toolkit/nifi-toolkit-tls/src/main/java/org/apache/nifi/test/util/FileUtils.java]: the osx specific nature of this is concerning

**Comment by joewitt** [comment]
I dont fully understand what this aims to achieve so please understand that in my comments.  However, changes to the nifi-toolkit-tls module dont seem to apply to wanting a test harness style capability for flows.  Adding a series of 'test' packages to main/source code seems like an anti-pattern.    We really need to reset on this effort and make sure it is clear what we're looking to achieve before this goes further as-is.

**Comment by peter-gergely-horvath** [comment]
Hi @joewitt and @MikeThomsen I messed up this pull request with some incorrectly merged code changes including the non-NiFi groupId and the whitespace change in nifi-toolkit-tls module etc. 
Sorry about that! For now, I will close any pull requests.

Since you had some really valuable ideas, could you please review the code in https://github.com/peter-gergely-horvath/nifi/tree/master/nifi-testharness and create issues with the inputs you have? I've implemented changes you mentioned in this pull request there already.

File: nifi-test/pom.xml (hunk 1 above)
Comment (peter-gergely-horvath) [reply nifi-test/pom.xml]: Yes, this was incorrect: I'll fix in my fork and create a new pull request later on.

File: nifi-test/pom.xml (hunk 2 above)
Comment (peter-gergely-horvath) [reply nifi-test/pom.xml]: It makes sense. I'll fix in my fork and create a new pull request later on.

File: nifi-docs/src/main/asciidoc/developer-guide.adoc (hunk 3 above)
Comment (peter-gergely-horvath) [reply nifi-docs/src/main/asciidoc/developer-guide.adoc]: The idea is that users should be able to write test cases in Java for NiFi flows: these code samples linked there would demonstrate creating a test case like that.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-documentation/src/main/java/org/apache/nifi/documentation/DocGenerator.java (hunk 4 above)
Comment (peter-gergely-horvath) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-documentation/src/main/java/org/apache/nifi/documentation/DocGenerator.java]: yes, indeed, this somehow slipped in: I'll fix in my fork and create a new pull request later on.

File: nifi-toolkit/nifi-toolkit-tls/src/main/java/org/apache/nifi/test/SimpleNiFiFlowDefinitionEditor.java (hunk 5 above)
Comment (peter-gergely-horvath) [reply nifi-toolkit/nifi-toolkit-tls/src/main/java/org/apache/nifi/test/SimpleNiFiFlowDefinitionEditor.java]: I'll fix in my fork and create a new pull request later on.

**Comment by peter-gergely-horvath** [comment]
@joewitt The idea is that this component would allow you to write Java test cases against your NiFi flow. Something like [NiFiFlowTest](https://github.com/peter-gergely-horvath/nifi/blob/master/nifi-testharness/src/test/java/org/apache/nifi/testharness/samples/NiFiFlowTest.java#L101).

The JavaDoc on [TestNiFiInstance](https://github.com/peter-gergely-horvath/nifi/blob/master/nifi-testharness/src/main/java/org/apache/nifi/testharness/TestNiFiInstance.java#L45) should explain the concept quite well.

**Pull request closed without merging**